src/main/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperation.java

```diff
@@ -84,25 +92,121 @@ public void updateUserAccess(CaseDetails caseDetails, CaseUser caseUser) {
+            orgNewUserCountMap.forEach((organisationId, newUserCount) ->
+                supplementaryDataRepository.incrementSupplementaryData(caseReference, "orgs_assigned_users." + organisationId, newUserCount)
+            )
+        );
+    }
+
+    private Map<Long, List<CaseAssignedUserRoleWithOrganisation>> getMapOfCaseAssignedUserRolesByCaseId(
+        List<CaseAssignedUserRoleWithOrganisation> caseUserRoles
+    ) {
+
+        Map<Long, List<CaseAssignedUserRoleWithOrganisation>> cauRolesByCaseId = new HashMap<>();
+
         List<Long> caseReferences = caseUserRoles.stream()
-            .map(CaseAssignedUserRole::getCaseDataId)
+            .map(CaseAssignedUserRoleWithOrganisation::getCaseDataId)
             .distinct()
             .map(Long::parseLong)
             .collect(Collectors.toCollection(ArrayList::new));
 
+        // create map of case references to case IDs
         Map<Long, String> caseIdAndReferences = getCaseDetailsList(caseReferences).stream()
             .collect(Collectors.toMap(CaseDetails::getReference, CaseDetails::getId));
 
-        caseUserRoles.forEach(caseUserRole -> {
-            final Long caseReference = Long.parseLong(caseUserRole.getCaseDataId());
+        // group roles by case reference
+        Map<String, List<CaseAssignedUserRoleWithOrganisation>> cauRolesByCaseReference = caseUserRoles.stream()
+            .collect(Collectors.groupingBy(CaseAssignedUserRoleWithOrganisation::getCaseDataId));
+
+        // merge both maps to check we have found all cases
+        cauRolesByCaseReference.forEach((key, roles) -> {
+            final Long caseReference = Long.parseLong(key);
             if (caseIdAndReferences.containsKey(caseReference)) {
                 final Long caseId = Long.parseLong(caseIdAndReferences.get(caseReference));
-                caseUserRepository.grantAccess(caseId, caseUserRole.getUserId(), caseUserRole.getCaseRole());
+                cauRolesByCaseId.put(caseId, roles);
             } else {
-                throw new CaseNotFoundException(caseReference.toString());
+                throw new CaseNotFoundException(key);
+            }
+        });
+
+        return cauRolesByCaseId;
+    }
+
+    private Map<String, Map<String, Object>> getNewUserCountByCaseAndOrganisation(Map<Long, List<CaseAssignedUserRoleWithOrganisation>> cauRolesByCaseId) {
+        Map<String, Map<String, Object>> result = new HashMap<>();
+
+        Map<Long, List<CaseAssignedUserRoleWithOrganisation>> caseUserRolesWhichHaveAnOrgId = cauRolesByCaseId.entrySet().stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().stream()
+                // filter out no organisation_id
+                .filter(caseUserRole -> StringUtils.isNoneBlank(caseUserRole.getOrganisationId()))
+                .collect(Collectors.toList())))
+            // filter cases that have no remaining roles
+            .entrySet().stream().filter(e -> !e.getValue().isEmpty())
```

Comment:
May be it's better to extract these into separate small methods for better readability and unit testing?

Reply:
I do agree. I'll add this to the content of the ticket for minor tweaks of this major work.